File: Rust/src/doubly_linked_list.rs
use std::{
    cell::RefCell,
    rc::{Rc, Weak},
};

#[derive(Debug)]
struct Node<T> {
    value: T,
    next: Option<Rc<RefCell<Node<T>>>>,
    prev: Option<Weak<RefCell<Node<T>>>>,
}

#[derive(Debug)]
pub struct Doubly_linked_list<T> {
    head: Option<Rc<RefCell<Node<T>>>>,
    tail: Option<Rc<RefCell<Node<T>>>>,
}

impl<T: std::fmt::Debug> Doubly_linked_list<T> {
    pub fn new() -> Self {
        Doubly_linked_list {
            head: None,
            tail: None,
        }
    }

    pub fn push_front(&mut self, value: T) {
        let new_node = Rc::new(RefCell::new(Node {
            value,
            next: self.head.clone(),
            prev: None,
        }));

        if let Some(old_head) = self.head.take() {
            old_head.borrow_mut().prev = Some(Rc::downgrade(&new_node));
        } else {
            self.tail = Some(new_node.clone());
        }

        self.head = Some(new_node);
    }

    pub fn pop_front(&mut self) -> Option<T> {
        self.head.take().map(|old_head| {
            self.head = old_head.borrow().next.clone();

            if let Some(new_head) = &self.head {
                new_head.borrow_mut().prev = None;
            } else {
                self.tail = None;
            }

            Rc::try_unwrap(old_head).ok().unwrap().into_inner().value
        })
    }

    pub fn print_list(&self) {
        let mut current = self.head.clone();
        print!("List: ");
        while let Some(node) = current {
            print!("{:?} <-> ", node.borrow().value);
            current = node.borrow().next.clone();
        }

        println!("None");
    }
}

File: go/list/d_linked_list/doubly_linked_list.go
package d_linked_list

import "fmt"

type Node struct {
	Next  *Node
	Value int
	Prev  *Node
}

type DoublyLinkedList struct {
	Head *Node
	Tail *Node
}

func (dl *DoublyLinkedList) Append(val int) {
	newNode := &Node{Value: val, Prev: nil, Next: nil}
	if dl.Head == nil {
		dl.Head = newNode
		dl.Tail = newNode
		return
	}
	newNode.Prev = dl.Tail
	dl.Tail.Next = newNode
	dl.Tail = newNode
}

func (dl *DoublyLinkedList) Prepend(val int) {
	newNode := &Node{Value: val, Prev: nil, Next: nil}

	if dl.Head == nil {
		dl.Head = newNode
		dl.Tail = newNode
		return
	}

	newNode.Next = dl.Head
	dl.Head.Prev = newNode
	dl.Head = newNode
}

func (dl *DoublyLinkedList) Delete(val int) {
	current := dl.Head

	for current != nil {
		if current.Value == val {
			if current == dl.Head && current == dl.Tail {
				dl.Head = nil
				dl.Tail = nil
				return
			}

			if current == dl.Head {
				dl.Head = current.Next
				dl.Head.Prev = nil
				return
			}

			if current == dl.Tail {
				dl.Tail = current.Prev
				dl.Tail.Prev = nil
				return
			}
			current.Prev.Next = current.Next
			current.Next.Prev = current.Prev
			return
		}
		current = current.Next
	}
}

func (dl *DoublyLinkedList) Display() {
	current := dl.Head

	for current != nil {
		fmt.Printf("%d <-> ", current.Value)
		current = current.Next
	}
	fmt.Println()
}

func (dl *DoublyLinkedList) DisplayReverse() {
	current := dl.Tail
	for current != nil {
		fmt.Printf("%d <-> ", current.Value)
		current = current.Prev
	}
	fmt.Println()
}
